vllm/model_executor/models/aria.py

```diff
@@ -0,0 +1,743 @@
+
+    Args:
+        patch_to_query_dict (dict): Maps patch numbers to their corresponding query numbers,
+            e.g., {1225: 128, 4900: 256}. This allows for different query sizes based on image resolution.
+        embed_dim (int): Embedding dimension.
+        num_heads (int): Number of attention heads.
+        kv_dim (int): Dimension of key and value.
+        ff_dim (int): Hidden dimension of the feed-forward network.
+        output_dim (int): Output dimension.
+        norm_layer (nn.Module): Normalization layer. Default is nn.LayerNorm.
+
+    Outputs:
+        A tensor with the shape of (batch_size, query_number, output_dim)
+    """
+
+    def __init__(
+        self,
+        patch_to_query_dict,
+        embed_dim,
+        num_heads,
+        kv_dim,
+        ff_dim,
+        output_dim,
+        norm_layer=nn.LayerNorm,
+    ):
+        super().__init__()
+        self.patch_to_query_dict = patch_to_query_dict
+        self.embed_dim = embed_dim
+        self.num_heads = num_heads
+
+        self.query = nn.Parameter(
+            torch.zeros(max(patch_to_query_dict.values()), self.embed_dim))
+
+        trunc_normal_(self.query, std=0.02)
+
+        self.cross_attn = CrossAttention(kv_dim, embed_dim, num_heads)
+
+        self.ln_ffn = norm_layer(embed_dim)
+        self.ffn = FFN(embed_dim, ff_dim, output_dim)
+
+        self.apply(self._init_weights)
+
+    def _init_weights(self, m):
+        if isinstance(m, nn.Linear):
+            trunc_normal_(m.weight, std=0.02)
+            if isinstance(m, nn.Linear) and m.bias is not None:
+                nn.init.constant_(m.bias, 0)
+        elif isinstance(m, nn.LayerNorm):
+            nn.init.constant_(m.bias, 0)
+            nn.init.constant_(m.weight, 1.0)
+
+    def forward(self, x, attn_mask=None):
+        """
+        Forward pass of the Projector module.
+
+        Args:
+            x (torch.Tensor): Input tensor of shape (batch_size, num_patches, kv_dim).
+            attn_mask (torch.Tensor, optional): Attention mask. Default is None.
+
+        Returns:
+            torch.Tensor: Output tensor of shape (batch_size, query_number, output_dim).
+        """
+        bs = x.shape[0]
+        queries = self.query.unsqueeze(0).repeat(bs, 1, 1)
+
+        query_num = self.patch_to_query_dict.get(x.shape[1], None)
+        assert (query_num is not None
+                ), f"Query number for {x.shape[1]} patches is not provided"
+
+        queries = queries[:, :query_num, :]
+
+        if attn_mask is not None:
+            attn_mask = attn_mask.repeat_interleave(self.num_heads, 0)
+            attn_mask = attn_mask.unsqueeze(1).expand(-1, queries.size(1), -1)
+
+        attention_out = self.cross_attn(x, queries, attn_mask=attn_mask)
+
+        out = self.ffn(self.ln_ffn(attention_out))
+
+        return out
+
+
+class Experts(nn.Module):
+
+    def __init__(self, config: AriaMoELMConfig):
```

Comment:
I think we can use the existing `FusedMoE` layer to replace this.

https://github.com/vllm-project/vllm/blob/f0e02380169b99a20cc5a4cd1848bbe085b50d5c/vllm/model_executor/layers/fused_moe/layer.py#L154-L175

Reply:
Sounds good, I'll give it a try!

Reply:
Done!

I'm going to battle with the formatter!

Reply:
You can directly run `bash ./format.sh` to address lints errors.

Reply:
Currently, my implementation based on `FusedMoE` only works with [`rhymes-ai/Aria-sequential_mlp`](https://huggingface.co/rhymes-ai/Aria-sequential_mlp). This model uses the same weights as [`rhymes-ai/Aria`](https://huggingface.co/rhymes-ai/Aria) but the weights are in a different format. After reconsidering, I believe it’s better to support `rhymes-ai/Aria`, which is the main repository on our Hugging Face Hub. I plan to modify the implementation accordingly.